Signal/src/textsecure/Messages/TSMessagesManager.m

```diff
@@ -179,6 +179,21 @@ - (void)handlePreKeyBundle:(IncomingPushMessageSignal*)preKeyMessage{
 }
 
 - (void)handleIncomingMessage:(IncomingPushMessageSignal*)incomingMessage withPushContent:(PushMessageContent*)content{
+    if(content.group!= nil )  {
+        __block BOOL ignoreMessage = NO;
+        [self.dbConnection readWithBlock:^(YapDatabaseReadTransaction *transaction) {
+            GroupModel *emptyModelToFillOutId = [[GroupModel alloc] initWithTitle:nil memberIds:nil image:nil groupId:content.group.id]; // TODO refactor the TSGroupThread to just take in an ID (as it is all that it uses). Should not take in more than it uses
+            TSGroupThread *gThread = [TSGroupThread threadWithGroupModel:emptyModelToFillOutId transaction:transaction];
+            if(gThread==nil) {
+                ignoreMessage = YES;
+            }
+        }];
+        if(ignoreMessage) {
+            DDLogDebug(@"recevied message from group that I left, ignoring");
```

Comment:
a) I don't think people appreciate stuff from a private channel being
posted (I don't) unless you remove stuff/ask first
b) sure, we can discuss. it makes sense to do the same thing android does.
sounds like we should have a sync up or read the code about that.
On Fri, Jan 2, 2015 at 9:13 PM, Frederic Jacobs <notifications@github.com>
wrote:

> In Signal/src/textsecure/Messages/TSMessagesManager.m
> <https://github.com/WhisperSystems/Signal-iOS/pull/248#discussion-diff-22423075>
> :
>
> > @@ -179,6 +179,21 @@ - (void)handlePreKeyBundle:(IncomingPushMessageSignal*)preKeyMessage{
> >  }
> >
> >  - (void)handleIncomingMessage:(IncomingPushMessageSignal*)incomingMessage withPushContent:(PushMessageContent*)content{
> > +    if(content.group!= nil )  {
> > +        __block BOOL ignoreMessage = NO;
> > +        [self.dbConnection readWithBlock:^(YapDatabaseReadTransaction *transaction) {
> > +            GroupModel *emptyModelToFillOutId = [[GroupModel alloc] initWithTitle:nil memberIds:nil image:nil groupId:content.group.id]; // TODO refactor the TSGroupThread to just take in an ID (as it is all that it uses). Should not take in more than it uses
> > +            TSGroupThread *gThread = [TSGroupThread threadWithGroupModel:emptyModelToFillOutId transaction:transaction];
> > +            if(gThread==nil) {
> > +                ignoreMessage = YES;
> > +            }
> > +        }];
> > +        if(ignoreMessage) {
> > +            DDLogDebug(@"recevied message from group that I left, ignoring");
>
> Ok, just read the backlog. That doesn't sound like a good idea.
>
> 1) Leaks metadata of who is in a group.
> 2) More things to keep in Redis
> 3) More bandwidth usage because those message will be sent out every time
> you connect to the socket and will eventually pile up over time.
>
> For the record:
> 13:20 < _corbett> no. so b) say I get a message from someone in a group
> that I have left (due to asynchronicity). does the UI display?
> 13:21 @lilia _corbett: oh i see
> 13:21 @lilia i think we ignore any group messages before the create
> 13:21 < jakeok> intuitively, i'd say don't display it
> 13:21 < _corbett> or a) say I get a message for a group prior to getting
> the group create message (due to asychronicity) do I wait for the latter?
> display immediately?
> 13:21 @lilia and send a leave-group to any messages received after the
> leave
> 13:21 < _corbett> wellll…. it could happen I get messages out of order
> 13:21 @lilia but odnt display them
> 13:21 < _corbett> so the person sent the message thinking I was still a
> member of the group
> 13:22 @lilia _corbett: you need to send them a leave group message until
> they stop
> 13:22 < jakeok> yeah i feel like the right way to do it technically is to
> keep the messages stored, and if you get an initial group context, you can
> display those messages in the conversation that you received prior to the
> create
> 13:22 @rhodey jakeok: ok, thanks for the heads up
> 13:22 < _corbett> yes… but that breaks our asychronicity too I think. we
> don’t wait typically
> 13:22 < _corbett> or try to impose order re a)
> 13:22 < _corbett> so I’m both wondering how android handles it, and if we
> want to handle it that way long term
> 13:23 @lilia i would not store rando group messages in the hope that the
> create is coming
> 13:23 < _corbett> for now I want to handle it exactly as android. I can
> just read the code.
> 13:23 < _corbett> hmmmm…. but what if there is just one create sent?
> 13:23 < _corbett> if you somehow missed it you will *never* get a message
> of the group
> 13:23 < _corbett> also breaks things I think.
> 13:24 @lilia the server should resend the group create until you 200OK it
> 13:24 < jakeok> unfortunately group messages are one of the few stateful
> things here
> 13:24 < _corbett> this is what I was unclear about and am asking about
> 13:24 @lilia thats how websocket resources work
> 13:24 < jlund> Still getting a lot of reports of unreliable message
> delivery, just to add more fodder to the current conversation.
> 13:24 @lilia if you dont ack the message with 200 OK response it will
> retry
> 13:24 < _corbett> still like the creator fires off the create and then 10
> messages and then some people respond. but for some reason on my phone it
> didn’t get the create
> 13:25 < _corbett> finally I 200OK it but in the meantime everyone in the
> group planned dinner thinking I was in the gorup
> 13:25 @lilia lol
> 13:25 < _corbett> I drop all those messages?
> 13:26 < jlund> jakeok: Do you know if the plan is to switch to using
> WebSockets for message delivery/retrieval on the Android side with GCM
> serving as a means of waking up the socket (similar to iOS)?
> 13:26 @lilia just dont 200OK the group messages, the server should have
> your back
> 13:26 < jakeok> jlund: that's my impression and hope
> 13:26 < jakeok> GCM is bad news bears
> 13:26 < jakeok> so we should just use it for signaling to connect
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/WhisperSystems/Signal-iOS/pull/248/files#r22423075>.
>



-- 
Dr. Corbett Moran
Rockets @ http://www.spacex.com
Physics @ http://www.ICS.uzh.ch  <http://www.ICS.uzh.ch>
Dev @ http://circleof6app.com
Dev @ https://whispersystems.org
www.christinecorbettmoran.com